cpp/src/hash/md5_hash.cu

```diff
@@ -27,23 +27,314 @@
+{
+  uint32_t const len  = sizeof(TKey);
+  uint8_t const* data = reinterpret_cast<uint8_t const*>(&key);
+  hash_state->message_length += len;
+
+  // 64 bytes are processed in each hash step
+  constexpr int md5_chunk_size = 64;
+  if (hash_state->buffer_length + len < md5_chunk_size) {
+    std::memcpy(hash_state->buffer + hash_state->buffer_length, data, len);
+    hash_state->buffer_length += len;
+  } else {
+    uint32_t copylen = md5_chunk_size - hash_state->buffer_length;
+
+    std::memcpy(hash_state->buffer + hash_state->buffer_length, data, copylen);
+    md5_hash_step(hash_state);
+
+    while (len > md5_chunk_size + copylen) {
+      std::memcpy(hash_state->buffer, data + copylen, md5_chunk_size);
+      md5_hash_step(hash_state);
+      copylen += md5_chunk_size;
+    }
+
+    std::memcpy(hash_state->buffer, data + copylen, len - copylen);
+    hash_state->buffer_length = len - copylen;
+  }
+}
 
 // MD5 supported leaf data type check
 bool md5_type_check(data_type dt)
 {
   return !is_chrono(dt) && (is_fixed_width(dt) || (dt.id() == type_id::STRING));
 }
 
-}  // namespace
+struct MD5ListHasher {
+  template <typename T, std::enable_if_t<is_chrono<T>()>* = nullptr>
+  void __device__ operator()(column_device_view data_col,
+                             size_type offset_begin,
+                             size_type offset_end,
+                             md5_intermediate_data* hash_state) const
+  {
+    cudf_assert(false && "MD5 Unsupported chrono type column");
+  }
 
-namespace detail {
+  template <typename T, std::enable_if_t<!is_fixed_width<T>()>* = nullptr>
+  void __device__ operator()(column_device_view data_col,
+                             size_type offset_begin,
+                             size_type offset_end,
+                             md5_intermediate_data* hash_state) const
+  {
+    cudf_assert(false && "MD5 Unsupported non-fixed-width type column");
+  }
+
+  template <typename T, std::enable_if_t<is_floating_point<T>()>* = nullptr>
+  void __device__ operator()(column_device_view data_col,
+                             size_type offset_begin,
+                             size_type offset_end,
+                             md5_intermediate_data* hash_state) const
+  {
+    for (int i = offset_begin; i < offset_end; i++) {
+      if (!data_col.is_null(i)) {
+        md5_process(normalize_nans_and_zeros_helper<T>(data_col.element<T>(i)), hash_state);
+      }
+    }
+  }
+
+  template <
+    typename T,
+    std::enable_if_t<is_fixed_width<T>() && !is_floating_point<T>() && !is_chrono<T>()>* = nullptr>
+  void CUDA_DEVICE_CALLABLE operator()(column_device_view data_col,
+                                       size_type offset_begin,
+                                       size_type offset_end,
+                                       md5_intermediate_data* hash_state) const
+  {
+    for (int i = offset_begin; i < offset_end; i++) {
```

Comment:
Can this operator take iterators rather than a view and indices?

Reply:
@harrism Is this a case where we would need to use an optional iterator or a pair iterator? I am not yet familiar with all the APIs for this task and their relative merits.

Reply:
This would be a nice use for an optional iterator. The optional iterator for column views returns optionals that contain a value if the row is not null. The pair iterator (which returns pairs where p.second is the validity flag) was the original tool for this purpose but is a bit less idiomatic than the optional, which is precisely designed to handle a possibly null value as opposed to a pair of arbitrary values (and is also more suitable for optimizations). In this case, [at the call site of the dispatch](https://github.com/rapidsai/cudf/pull/9212/files/03bfe6d30aea7bb591436c1e30ae82c823d6cb3f#diff-420ce2d2376712ffb08f35d23344c1e01b4eead33aa107ef63d555251bc20929R322) you could do something like
```
auto start = data.optional_begin<T>(cudf::contains_nulls::YES{}) + offsets.element<size_type>(row_index);
auto end = data.optional_begin<T>(cudf::contains_nulls::YES{}) + offsets.element<size_type>(row_index + 1);
```
and then change this operator to accept the two iterators rather than the data and the indices. You could then check validity using the optional's `has_value` method.

Reply:
Unfortunately it does not appear that we can use an optional iterator as an argument to `operator()` because we have to perform a type dispatch to get `T`. However, it is possible to have a catch-all `operator()` that then dispatches the iterator. I think that is still cleaner and allows me to centralize the logic handling element validity.